plugins/ai/copilot-chat.nix

```diff
@@ -0,0 +1,316 @@
+
+            kind = helpers.mkNullOrStr "Kind of this prompt.";
+
+            mapping = helpers.mkNullOrStr "A key to bind to this prompt.";
+
+            system_prompt = helpers.mkNullOrStr "System prompt to use.";
+
+            callback = helpers.mkNullOrLuaFn ''
+              Callback to trigger when this prompt is executed.
+              `fun(response: string, source: CopilotChat.config.source)`
+            '';
+
+            selection = helpers.mkNullOrLuaFn ''
+              Selection for this prompt.
+              `fun(source: CopilotChat.config.source):CopilotChat.config.selection`
+            '';
+          };
+        };
+
+        default = {
+          Explain.prompt = "/COPILOT_EXPLAIN Write an explanation for the active selection as paragraphs of text.";
+          Review = {
+            prompt = "/COPILOT_REVIEW Review the selected code.";
+            callback = ''
+              function(response, source)
+                -- see config.lua for implementation
+              end
+            '';
+          };
+          Fix.prompt = "/COPILOT_GENERATE There is a problem in this code. Rewrite the code to show it with the bug fixed.";
+          Optimize.prompt = "/COPILOT_GENERATE Optimize the selected code to improve performance and readablilty.";
+          Docs.prompt = "/COPILOT_GENERATE Please add documentation comment for the selection.";
+          Tests.prompt = "/COPILOT_GENERATE Please generate tests for my code.";
+          FixDiagnostic = {
+            prompt = "Please assist with the following diagnostic issue in file:";
+            selection = "require('CopilotChat.select').diagnostics";
+          };
+          Commit = {
+            prompt = "Write commit message for the change with commitizen convention. Make sure the title has maximum 50 characters and message is wrapped at 72 characters. Wrap the whole message in code block with language gitcommit.";
+            selection = "require('CopilotChat.select').gitdiff";
+          };
+          CommitStaged = {
+            prompt = "Write commit message for the change with commitizen convention. Make sure the title has maximum 50 characters and message is wrapped at 72 characters. Wrap the whole message in code block with language gitcommit.";
+            selection = ''
+              function(source)
+                return select.gitdiff(source, true)
+              end
+            '';
+          };
+        };
+      in
+      helpers.defaultNullOpts.mkAttrsOf (with types; either str promptType) default ''
+        Default prompts.
+      '';
+
+    window = {
+      layout =
+        helpers.defaultNullOpts.mkEnumFirstDefault
+          [
+            "vertical"
+            "horizontal"
+            "float"
+            "replace"
+          ]
+          ''
+            Layout for the window.
+          '';
+
+      width =
+        helpers.defaultNullOpts.mkNullable (with types; either (number.between 0.0 1.0) ints.positive) 0.5
```

Comment:
Should this (and other similar options) accept a raw type too?

Reply:
Maybe we should be reaching for `mkNullableWithRaw` for most custom-type options? 🤔

Reply:
Yes, good idea. Done.